### PR TITLE
Issue #2710985 by mglaman: Add an order summary

### PR DESCRIPTION
**Pull request by mglaman**
order_information___site-install

### CODE REVIEW

File: modules/checkout/templates/commerce-checkout-form.html.twig [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * @ingroup themeable
+ */
+#}
+<div class="layout-checkout-form clearfix">
+  <div class="layout-region layout-region-checkout-main">
+    {{ form|without('cart_summary', 'actions') }}
+  </div>
+  <div class="layout-region layout-region-checkout-secondary">
+    <h3>{% trans %} Order Summary {% endtrans %}</h3>
```
Comment (bojanz) [inline modules/checkout/templates/commerce-checkout-form.html.twig]: Can the Order Summary title not come from the form element that's returned by buildCartSummary()?

File: modules/checkout/src/Plugin/Commerce/CheckoutFlow/MultistepDefault.php [hunk 2]
```diff
@@ -20,13 +20,16 @@ public function getSteps() {
...
+        'has_cart_summary' => FALSE,
       ],
       'order_information' => [
         'label' => $this->t('Order information'),
+        'has_cart_summary' => TRUE,
       ],
       'review' => [
         'label' => $this->t('Review'),
         'next_label' => $this->t('Continue to review'),
+        'has_cart_summary' => FALSE,
```
Comment (bojanz) [inline modules/checkout/src/Plugin/Commerce/CheckoutFlow/MultistepDefault.php]: let's have it TRUE for review for now. 
We'd make it FALSE if we provide a different summary for the review page alone.

File: modules/checkout/templates/commerce-checkout-form.html.twig (hunk 1 above)
Comment (mglaman) [inline modules/checkout/templates/commerce-checkout-form.html.twig]: When you embed a view via theme or other it does not render the view title. For profile we had to make the following hack

```
        '#pre_render' => [
          ['\Drupal\views\Element\View', 'preRenderViewElement'],
          'profile_views_add_title_pre_render',
        ],
```

which adds it

File: modules/checkout/src/Plugin/Commerce/CheckoutFlow/CheckoutFlowBase.php [hunk 3]
```diff
@@ -166,6 +180,22 @@ public function getVisibleSteps() {
   /**
    * {@inheritdoc}
    */
+  public function buildCartSummary(array $form, FormStateInterface $form_state) {
+    $summary_view_id = $this->configuration['cart_summary_view'];
```
Comment (bojanz) [inline modules/checkout/src/Plugin/Commerce/CheckoutFlow/CheckoutFlowBase.php]: No point in declaring the $summary_view_id variable if it's only used once in the returned array?

File: modules/checkout/templates/commerce-checkout-form.html.twig (hunk 1 above)
Comment (mglaman) [inline modules/checkout/templates/commerce-checkout-form.html.twig]: I thought about adding it as a view header, but we can't guarantee the `basic_html` filter is present :/

File: modules/checkout/config/install/core.entity_view_display.commerce_order.default.summary.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+status: true
+dependencies:
+  config:
+    - commerce_order.commerce_order_type.default
+    - core.entity_view_mode.commerce_order.summary
+    - field.field.commerce_order.default.line_items
+  module:
+    - commerce_price
+_core:
+  default_config_hash: '-_kLkKivfKWGL7kEtXmU6JIMhj3mwG8QhYrD4_kywD0'
```
Comment (bojanz) [inline modules/checkout/config/install/core.entity_view_display.commerce_order.default.summary.yml]: We don't want the default hash.

File: modules/checkout/config/install/views.view.commerce_checkout_cart_summary.yml [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+status: true
+dependencies:
+  enforced:
+    module:
+      - commerce_cart
+  module:
+    - commerce_order
+    - commerce_price
+_core:
+  default_config_hash: uNc9tmnfjDsZtEjCgwSe1cb8uGPKOsjnzPrRdYUZ28g
```
Comment (bojanz) [inline modules/checkout/config/install/views.view.commerce_checkout_cart_summary.yml]: We don't want the default hash.

**Comment by mglaman** [comment]
login___site-install_and_order_information___site-install

File: modules/checkout/commerce_checkout.info.yml [hunk 6]
```diff
@@ -10,3 +10,6 @@ dependencies:
 config_devel:
   install:
     - commerce_checkout.commerce_checkout_flow.default
+    - core.entity_view_display.commerce_order.default.summary
+    - core.entity_view_mode.commerce_order.summary
+    - views.view.commerce_checkout_cart_summary
```
Comment (bojanz) [inline modules/checkout/commerce_checkout.info.yml]: views.view.commerce_checkout_order_summary

File: modules/checkout/src/Plugin/Commerce/CheckoutFlow/CheckoutFlowInterface.php [hunk 7]
```diff
@@ -79,4 +80,17 @@ public function getSteps();
    */
   public function getVisibleSteps();
 
+  /**
+   * Builds an order summary for the checkout step.
```
Comment (bojanz) [inline modules/checkout/src/Plugin/Commerce/CheckoutFlow/CheckoutFlowInterface.php]: Builds the order summary for the checkout step.

File: modules/checkout/src/Plugin/Commerce/CheckoutFlow/CheckoutFlowBase.php [hunk 8]
```diff
@@ -54,12 +62,15 @@
    *   The event dispatcher.
    * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
    *   The route match.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager.
    */
-  public function __construct(array $configuration, $plugin_id, $plugin_definition, EventDispatcherInterface $event_dispatcher, RouteMatchInterface $route_match) {
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, EventDispatcherInterface $event_dispatcher, RouteMatchInterface $route_match, EntityTypeManagerInterface $entity_type_manager) {
```
Comment (bojanz) [inline modules/checkout/src/Plugin/Commerce/CheckoutFlow/CheckoutFlowBase.php]: Let's inject the entity type manager before the event dispatcher, it's more essential (and it's alphabetic),

File: modules/checkout/templates/commerce-checkout-form.html.twig [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+ * Two column template for the checkout form.
+ *
+ * Available variables:
+ * - form: The form.
+ *
+ * @ingroup themeable
+ */
+#}
+<div class="layout-checkout-form clearfix">
+    {{ form }}
```
Comment (bojanz) [inline modules/checkout/templates/commerce-checkout-form.html.twig]: Wrong indentation.

**Comment by mglaman** [comment]
@bojanz the last item was a "summary" view mode. Right now the view uses fields and renders the quantity, product title, and product price in a field. You want to add the summary view mode to commerce_product_variation and have it render that entity in the view mode. So we'd just render Quantity field then Rendered Entity for the purchased entity as "summary."

Correct?

**Comment by bojanz** [comment]
@mglaman Correct

**Comment by mglaman** [comment]
order_information___site-install

@bojanz here is updated w/ last -f push that is running

**Comment by bojanz** [comment]
There was no need to move the Price to the main column, let's keep it in its own, like it was before, and like it is in the cart. All I want is the cart behavior.

**Comment by mglaman** [comment]
@bojanz I'm silly. I'll have it updated soon

**Comment by mglaman** [comment]
order_information___site-install

Next round.

File: modules/checkout/commerce_checkout.info.yml [hunk 10]
```diff
@@ -10,3 +10,8 @@ dependencies:
 config_devel:
   install:
     - commerce_checkout.commerce_checkout_flow.default
+    - core.entity_view_display.commerce_order.default.summary
```
Comment (bojanz) [inline modules/checkout/commerce_checkout.info.yml]: We don't need the "summary" view mode / form display for orders anymore? Won't be used?

File: modules/checkout/commerce_checkout.module [hunk 11]
```diff
@@ -19,12 +19,35 @@ function commerce_checkout_theme() {
         'steps' => [],
       ],
     ],
+    'commerce_checkout_form' => [
+      'render element' => 'form',
+    ],
+    'commerce_checkout_form__has_summary' => [
```
Comment (bojanz) [inline modules/checkout/commerce_checkout.module]: commerce_checkout_form__with_summary

File: modules/checkout/templates/commerce-checkout-form--has-summary.html.twig [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+ * @file
+ * Two column template for the checkout form.
+ *
+ * Available variables:
+ * - form: The form.
+ *
+ * @ingroup themeable
+ */
+#}
+{{ attach_library('commerce_checkout/form') }}
```
Comment (bojanz) [inline modules/checkout/templates/commerce-checkout-form--has-summary.html.twig]: The library is already attached by the form?

File: modules/checkout/commerce_checkout.info.yml (hunk 10 above)
Comment (mglaman) [inline modules/checkout/commerce_checkout.info.yml]: Shoot; I forgot our discussion - are we porting the order total footer here? or removing, then porting and adding in once this is closed a a follow up

Comment (bojanz) [inline modules/checkout/commerce_checkout.info.yml]: Removing, then adding it in a followup to both cart & order summary once the new area plugin is done.

**Comment by mglaman** [comment]
@bojanz finally removed.

**Comment by mglaman** [comment]
Created follow up for order total area https://www.drupal.org/node/2735441

**Comment by bojanz** [comment]
Committed in ed154b7db21df8e8c678c6e6a197563bab7f65c8.

Made it possible to disable the order summary by selecting None as the view value.

**Pull request closed without merging**